tutorialtool/src/state/appStateContext.tsx

```diff
@@ -0,0 +1,62 @@
+import React, { createContext, useEffect, useReducer } from "react";
+import { AppState, initialAppState } from "./state";
+import { Action } from "./actions";
+import reducer from "./reducer";
+import assert from "assert";
+
+let state: AppState;
+let dispatch: React.Dispatch<Action>;
+
+let initializationComplete: () => void;
+
+// This promise will resolve when the app state is initialized and available outside the React context.
+export const AppStateReady: Promise<boolean> = new Promise(resolve => {
+    initializationComplete = () => resolve(true);
+});
+
+// Never cache `state` and `dispatch`. They can and will be updated frequently.
+export const stateAndDispatch = () => {
```

Comment:
I actually ran into startup problems in teacher tool with this defined as an arrow function. Expressing is as a regular function fixed it (`export function stateAndDispatch(...)`). If you get bizarre invalid symbols referencing `stateAndDispatch` on app load, make this a regular function to fix it.